src/testing/slim_check/sampleable.lean

```diff
@@ -0,0 +1,280 @@
+dividing by 2, subtracting the result from `n` and alternating the signs.
+For example, `int.shrink 40 = [20, -20, 30, -30, 35, -35, 38, -38, 39, -39]`. -/
+def int.shrink (i : ℤ) : list ℤ :=
+int.shrink' (int.nat_abs i) (int.nat_abs i) []
+
+instance sampleable_int : sampleable ℤ :=
+{ sample := sized $ λ sz,
+       let k := sz^5 in
+       (λ n : fin _, n.val - int.of_nat (k / 2) ) <$> choose_any (fin $ succ k),
+  shrink := lazy_list.of_list ∘ int.shrink   }
+
+instance sampleable_bool : sampleable bool :=
+{ sample := do { x ← choose_any bool,
+                 return x },
+  shrink := λ _, lazy_list.nil }
+
+instance sampleable_prod {β} [sampleable α] [sampleable β] : sampleable (α × β) :=
+{ sample := do { ⟨x⟩ ← uliftable.up $ sample α,
+                 ⟨y⟩ ← uliftable.up $ sample β,
+                 pure (x,y) },
+  shrink := λ x, lazy_list.lseq prod.mk (shrink x.1) (shrink x.2) }
+
+/-- shrinking function for sum types -/
+def sum.shrink {β} [sampleable α] [sampleable β] : α ⊕ β → lazy_list (α ⊕ β)
+| (sum.inr x) := (shrink x).map sum.inr
+| (sum.inl x) := (shrink x).map sum.inl
+
+instance sampleable_sum {β} [sampleable α] [sampleable β] : sampleable (α ⊕ β) :=
+{ sample := uliftable.up_map sum.inl (sample α) <|>
+            uliftable.up_map sum.inr (sample β),
+  shrink := sum.shrink _ }
+
+/-- `sampleable_char` can be specialized into customized `sampleable char` instances.
+
+The resulting instance has `1 / length` chances of making an unrestricted choice of characters
+and it otherwise chooses a character from `characters` with uniform probabilities.  -/
+def sampleable_char (length : nat) (characters : string) : sampleable char :=
+{ sample := do { x ← choose_nat 0 length dec_trivial,
+                 if x.val = 0 then do
+                   n ← sample ℕ,
+                   pure $ char.of_nat n
+                 else do
+                   i ← choose_nat 0 (characters.length - 1) dec_trivial,
+                   pure (characters.mk_iterator.nextn i).curr },
+  shrink := λ _, lazy_list.nil }
+
+instance : sampleable char :=
+sampleable_char 3 " 0123abcABC:,;`\\/"
+
+variables {α}
+
+/-- implementation of `sampleable (list α)` -/
+def list.shrink' (shrink_a : α → lazy_list α) : list α → lazy_list (list α)
+| [] := lazy_list.nil
+| (x :: xs) :=
+  let ys := list.shrink' xs in
+  interleave ys $ lazy_list.lseq (::) ((shrink_a x).append (lazy_list.singleton x)) (lazy_list.cons [] ys)
+
+/-- `list.shrink_with shrink_f xs` shrinks `xs` by deleting various items of the list
+and shrinking others (using `shrink_f`). `lseq` is being used to interleave the
+resulting shrunken lists to put smaller lists earlier in the results. -/
+def list.shrink_with (shrink_a : α → lazy_list α) (xs : list α) : lazy_list (list α) :=
+(list.shrink' shrink_a xs).init
+
+instance sampleable_list [sampleable α] : sampleable (list α) :=
+{ sample := list_of (sample α),
+  shrink := list.shrink_with shrink  }
+
+instance sampleable_prop : sampleable Prop :=
```

Comment:
What is the point of this instance? (I'm also confused why there are proofs in this file when everything here should be `meta` code.)

Reply:
Why would you make it into meta code?

Reply:
So that the proofs are not needed, the number of imports goes down, and the code gets simpler. Isn't that reason enough?

Reply:
I think the proofs are useful as a sanity check.